### PR TITLE
Fix time-series filtering

### LINKED ISSUE #202: Timestamp filtering returns empty DataFrame
## What went wrong?
Since version 0.3.x we are able to index `Timestamp` and `Date`. But when we try to filter but those Timestamps, the Query is not well parsed and returns an empty DataFrame.

After a few prints and tests, I discovered that the `d.getTime` we use in `LinearTransformer` to initialize `min` and `max` of `LinearTransformation`, returns a Long X. In contrast, the filtering from the user returns a different Long Y. 

This difference does not allow the `QuerySpecBuilder` to initialize the correct space. `QuerySpaceFromTo` returns an `EmptySpace`, which contains 0 files to read. 

### 1. Code that triggered the bug, or steps to reproduce:

Using some code in the `TransformerIndexingTest`:

```scala
    import spark.implicits._
    val df =
      Seq(
        "2017-01-03 12:02:00",
        "2017-01-02 12:02:00",
        "2017-01-02 12:02:00",
        "2017-01-02 12:02:00",
        "2017-01-01 12:02:00",
        "2017-01-01 12:02:00")
        .toDF("date")
        .withColumn("my_date", to_timestamp($"date"))

    df.write.format("qbeast").option("columnsToIndex", "my_date").save(tmpDir)

    val indexed = spark.read.format("qbeast").load(tmpDir)

    indexed.filter("my_date == '2017-01-02 12:02:00'").count() shouldBe 3
```

But the returned DataFrame contains 0 rows.

### 2. Branch and commit id:

I'm using photon branch: `photon-datasource-standalone`

### 3. Spark version:
On the spark shell run `spark.version`.

Spark version 3.3.0 and Delta version 2.1.0

### 4. Hadoop version:
On the spark shell run `org.apache.hadoop.util.VersionInfo.getVersion()`.

3.3.4

### 5. How are you running Spark?
Are you running Spark inside a container? Are you launching the app on a remote K8s cluster? Or are you just running the tests in a local computer?

Local

### PR DESCRIPTION
**Pull request by osopardo1**
## Description

Fixes #202 (and #201 😃 )

## Type of change

It is a Bug Fix for the reading part, it does not include breaking changes on the algorithm or the writing of the files.

Now `TimestampType` and `DateType` are pre-processed on `QueryFilterUtils` to return a Millisecond representation that can be filtered properly by the index. 

## Checklist:

Here is the list of things you should do before submitting this pull request:

- [x] New feature / bug fix has been committed following the [Contribution guide](https://github.com/Qbeast-io/qbeast-spark/blob/main/CONTRIBUTING.md).
- [x] Add comments to the code (make it easier for the community!).
- [x] Change the documentation.
- [x] Add tests.
- [x] Your branch is updated to the main branch (dependent changes have been merged).

## How Has This Been Tested? (Optional)

I've added a test on `src/test/scala/io/qbeast/spark/index/query` called `TimeSeriesQueryTest`. In this class, I've tested four different ways of creating an index on a Date / Timestamp column. Each one were failing differently, and should fail if executed with previous commits. 

**Test Configuration**:
 TBD.

### CODE REVIEW

**Comment by osopardo1** [comment]
To have a fixed version for 0.3.x for compatibility with Spark 3.2.x, the commits of this PR should be merged into a new 0.3.5 version release, and updated accordingly.

**Comment by osopardo1** [comment]
Hello @alexeiakimov ! Could you review this quick fix? Thank you 🙌

File: src/main/scala/io/qbeast/spark/index/query/QueryFiltersUtils.scala [hunk 1]
```diff
@@ -91,15 +96,25 @@ private[query] trait QueryFiltersUtils {
   }
 
   /**
-   * Convert an Spark String type to a Scala core type
-   * @param value the value to convert
+   * Convert an Literal value from Spark to a Qbeast/Scala core type
```
Comment (alexeiakimov) [inline src/main/scala/io/qbeast/spark/index/query/QueryFiltersUtils.scala:99]: Probably typo: a literal value

File: src/main/scala/io/qbeast/spark/index/query/QueryFiltersUtils.scala [hunk 2]
```diff
@@ -91,15 +96,25 @@ private[query] trait QueryFiltersUtils {
...
-  def sparkTypeToCoreType(value: Any): Any = {
-    value match {
-      case s: UTF8String => s.toString
-      case _ => value
+  def sparkTypeToCoreType(l: Literal): Any = {
+
+    (l.value, l.dataType) match {
+      case (int: Integer, _: DateType) =>
+        // convert DateType to Milliseconds
+        lazy val zoneId = getZoneId(SQLConf.get.sessionLocalTimeZone)
```
Comment (alexeiakimov) [inline src/main/scala/io/qbeast/spark/index/query/QueryFiltersUtils.scala:109]: I wonder that the data is always stored as the Spark session local timezone, right?

File: src/main/scala/io/qbeast/spark/index/query/QueryFiltersUtils.scala [hunk 3]
```diff
@@ -91,15 +96,25 @@ private[query] trait QueryFiltersUtils {
...
+        val dateInMicros = daysToMicros(int, zoneId)
+        val dateInMillis = TimeUnit.MILLISECONDS.convert(dateInMicros, TimeUnit.MICROSECONDS)
```
Comment (alexeiakimov) [inline src/main/scala/io/qbeast/spark/index/query/QueryFiltersUtils.scala:111]: Can be shorter:
```Scala
TimeUnit.MICROSECONDS.toMillis(dateInMicros)
```

File: src/main/scala/io/qbeast/spark/index/query/QueryFiltersUtils.scala [hunk 4]
```diff
@@ -91,15 +96,25 @@ private[query] trait QueryFiltersUtils {
...
+        dateInMillis
+      case (long: Long, _: TimestampType) =>
+        // convert Timestamp from Microseconds to Milliseconds
+        TimeUnit.MILLISECONDS.convert(long, TimeUnit.MICROSECONDS)
```
Comment (alexeiakimov) [inline src/main/scala/io/qbeast/spark/index/query/QueryFiltersUtils.scala:115]: This also can be shorter.

File: src/main/scala/io/qbeast/spark/index/query/QueryFiltersUtils.scala (hunk 2 above)
Comment (alexeiakimov) [reply src/main/scala/io/qbeast/spark/index/query/QueryFiltersUtils.scala:109]: Do you mean that the data is persisted on disk using the timezone of Spark?

File: src/main/scala/io/qbeast/spark/index/query/QueryFiltersUtils.scala (hunk 3 above)
Comment (osopardo1) [reply src/main/scala/io/qbeast/spark/index/query/QueryFiltersUtils.scala]: Thanks! Done 👍

File: src/main/scala/io/qbeast/spark/index/query/QueryFiltersUtils.scala (hunk 2 above)
Comment (osopardo1) [reply src/main/scala/io/qbeast/spark/index/query/QueryFiltersUtils.scala:109]: According to this post: https://www.databricks.com/blog/2020/07/22/a-comprehensive-look-at-dates-and-timestamps-in-apache-spark-3-0.html If no Timezone is specified, **Spark takes a time zone from the SQL configuration `spark.sql.session.timeZone` and applies it to function invocations**. 

So yes, data would be persisted on the Disk using the `spark.sql.session.timeZone`.

Comment (alexeiakimov) [reply src/main/scala/io/qbeast/spark/index/query/QueryFiltersUtils.scala:109]: My bad, after reading the article I think the code is correct.

**Review by alexeiakimov — APPROVED** [review]

**Comment by osopardo1** [comment]
Fixed #202 and #201

**Pull request merged**